File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head==null) return null;
        List<Integer> l=new ArrayList<>();
        ListNode temp=head;
        while(temp!=null){
            l.add(temp.val);
            temp=temp.next;
        }
        rotate(l,k);
        ListNode op=new ListNode(l.get(0));
        ListNode lp=op;
        for(int i=1;i<l.size();i++){
            ListNode lo=new ListNode(l.get(i));
            lp.next=lo;
            lp=lp.next;
        }

        return op;
    }
    public static void rotate(List<Integer> l, int k){
        // if(k==l.size()){
        //     rev(l,0,l.size()-1);
        // }
        int rot=k%l.size();
        rev(l,0,l.size()-rot-1);
        rev(l,l.size()-rot,l.size()-1);
        rev(l,0,l.size()-1);
    }
    public static void rev(List<Integer> l,int i, int j){
        while(i<j){
            int t=l.get(i);
            l.set(i,l.get(j));
            l.set(j,t);
            i++;
            j--;
        }
    }
}

File: 0061-rotate-list/0061-rotate-list.cs
// using System;
// using System.Collections.Generic;

// public class ListNode {
//     public int val;
//     public ListNode next;
//     public ListNode(int val = 0, ListNode next = null) {
//         this.val = val;
//         this.next = next;
//     }
// }

public class Solution {
    public ListNode RotateRight(ListNode head, int k) {
        if (head == null) return null;

        List<int> l = new List<int>();
        ListNode temp = head;

        while (temp != null) {
            l.Add(temp.val);
            temp = temp.next;
        }

        op(l, k);

        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        foreach (int val in l) {
            curr.next = new ListNode(val);
            curr = curr.next;
        }

        return dummy.next;
    }

    public static void op(List<int> l, int k) {
        k = k % l.Count;
        rev(0, l.Count - 1, l);
        rev(0, k - 1, l);
        rev(k, l.Count - 1, l);
    }

    public static void rev(int i, int j, List<int> l) {
        while (i <= j) {
            int t = l[i];
            l[i] = l[j];
            l[j] = t;
            i++;
            j--;
        }
    }
}
